### PR TITLE
Address, NodeId, ModuleId, SubstateKey, SubstateValue and Track Overhaul

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

- HRP for every stored entity, including internal vaults
- Redefined `GlobalAddress`, `LocalAddress`, `PackageAddress`, `ComponentAddress` and `ResourceAddress`;
- Cleaned up `Own` and `Reference` in value model
- Refactored `Track`/`Database` to be fully generic (You can now freely traverse the whole substate database)
- Refactored `NodeId`, `ModuleId` and `SubstateKey` to be fully generic
- Cleaned up  `CallFrame` implementation
- Removed `RuntimeSubstate` and `PersistedSubstate`, in favour of `IndexedScryptoValue`

## Limitations

- There is noticeable performance loss as a result of substate generalisation, for which we will optimise at SBOR layer. 
- State dependency tracking is not yet implemented (conflicting with iterable KVStore design)

### For Dapp Developers

You can now use `Address("internal_vault_xxxx")` for recall.

### For Internal Integrators

Pretty much every identifier is changed, so talk to Yulong if you have trouble with integration.

### CODE REVIEW

File: radix-engine-interface/src/constants.rs [hunk 1]
```diff
@@ -1,37 +1,40 @@
...
-    vanity_address!(EntityType::NonFungibleResource, 6);
+    resource_address(EntityType::GlobalNonFungibleResource, 6);
 pub const IDENTITY_OWNER_TOKEN: ResourceAddress =
-    vanity_address!(EntityType::NonFungibleResource, 7);
+    resource_address(EntityType::GlobalNonFungibleResource, 7);
 pub const ACCOUNT_OWNER_TOKEN: ResourceAddress =
-    vanity_address!(EntityType::NonFungibleResource, 8);
+    resource_address(EntityType::GlobalNonFungibleResource, 8);
 
 /// The address of the faucet package.
```
Comment (talekhinezh) [inline radix-engine-interface/src/constants.rs:25]: Lol invalid comment

File: radix-engine-interface/src/constants.rs [hunk 2]
```diff
@@ -52,39 +55,8 @@ pub fn is_native_package(address: PackageAddress) -> bool {
     }
 }
 
-pub const FAUCET_PACKAGE: PackageAddress = vanity_address!(EntityType::Package, 64);
+pub const FAUCET_PACKAGE: PackageAddress = package_address(EntityType::GlobalPackage, 64);
 pub const FAUCET_BLUEPRINT: &str = "Faucet";
 
-/// The address of the faucet component, test network only.
-pub const FAUCET_COMPONENT: ComponentAddress = construct_address!(
```
Comment (talekhinezh) [inline radix-engine-interface/src/constants.rs:59]: Note to self: find how we retrieve faucet component address

File: radix-engine-interface/src/types/node_layout.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    AccessRules1, // TODO: remove
+}
+
+impl Into<ModuleId> for TypedModuleId {
+    fn into(self) -> ModuleId {
+        ModuleId(self as u8)
+    }
+}
+
+#[repr(u8)]
```
Comment (talekhinezh) [inline radix-engine-interface/src/types/node_layout.rs:40]: fun!

File: radix-engine-stores/src/interface.rs [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+///
+/// In practice, we will likely end up with only one implementation.
+///
+/// The trait here is for formalizing the interface and intended user flow.
+pub trait SubstateStore {
+    /// Acquires a lock over a substate.
+    ///
+    /// # Panics
+    /// - If the module ID is invalid
+    fn acquire_lock(
```
Comment (talekhinezh) [inline radix-engine-stores/src/interface.rs:74]: I'm wondering if we even need locks at this layer. It was initially implemented as a sanity check.

File: radix-engine-common/src/data/scrypto/custom_schema.rs [hunk 5]
```diff
@@ -10,7 +10,9 @@ pub type ScryptoTypeData<L> = TypeData<ScryptoCustomTypeKind, L>;
 /// A schema for the values that a codec can decode / views as valid
 #[derive(Debug, Clone, PartialEq, Eq, Sbor)]
 pub enum ScryptoCustomTypeKind {
-    Address, /* any */
+    Reference, /* any */
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine-common/src/data/scrypto/custom_schema.rs:13]: Yes, they're `LocalAddress` vs `GlobalAddress`.  Reference type check will be the same between `Own` and `Reference`, by schema.

File: radix-engine-common/src/address/hrpset.rs [hunk 6]
```diff
@@ -1,41 +1,48 @@
...
-    access_controller_component: String,
+    component: String,
+    account: String,
+    identity: String,
+    epoch_manager: String,
+    clock: String,
+    validator: String,
+    access_controller: String,
+    internal_vault: String,
+    internal_access_controller: String,
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine-common/src/address/hrpset.rs:18]: It was listed there before, but I don't mind removing it :)

File: radix-engine-interface/src/constants.rs (hunk 1 above)
Comment (iamyulong) [reply radix-engine-interface/src/constants.rs:25]: Nice catch!

**Review by talekhinezh — APPROVED** [review]

**Pull request merged**